chore: add agentsocket api client

SasSwart opened the pull request:
relates to: https://github.com/coder/internal/issues/1094

This is number 4 of 5 pull requests in an effort to add agent script ordering. It adds an API client that can call the workspace agent socket API for script ordering operations.

In a follow-up PR, CLI commands will be added that consume this API.

I used an LLM to produce some of these changes, but I have conducted thorough self review and consider this contribution to be ready for an external reviewer.

SasSwart committed:
feat(codersdk): add socket client for agent communication

SasSwart committed:
make lint/go

SasSwart committed:
clean up agent socket client.

remove unused types and simplify the interface

SasSwart committed:
chore: use the new agentsdk socket client in server tests to prove it works

SasSwart committed:
Remove defunct testing client for the agentsocket API

SasSwart committed:
remove defunct testing client from the agentsocket api

SasSwart committed:
refine agent socket tests

mafredri reviewed (commented):
Since logic doesn't change much, this PR would normally be an easy approve, but fundamentally I think this might be the wrong approach. I think there should be a larger benefit than slight improvements in ergonomics of using the client as the burden to maintain it becomes greater. Are there any future plans I'm overlooking?

SasSwart committed:
Refine documentation for the agentsocket package

SasSwart commented on agent/agentsocket/client.go:
```diff
@@ -0,0 +1,137 @@
+		return nil, xerrors.Errorf("connect to socket: %w", err)
+	}
+
+	// Create drpc connection using the multiplexed connection
```
Done in [2b5b383](https://github.com/coder/coder/pull/20718/commits/2b5b383cf968e4944ac28e764506bef0f943cb90)

SasSwart commented on agent/agentsocket/service_test.go:
```diff
@@ -70,6 +67,8 @@ func TestDRPCAgentSocketService(t *testing.T) {
 
 		socketPath := filepath.Join(tempDirUnixSocket(t), "test.sock")
+		ctx, cancel := context.WithTimeout(context.Background(), testutil.WaitLong)
+		t.Cleanup(cancel)
```
Done in [2b5b383](https://github.com/coder/coder/pull/20718/commits/2b5b383cf968e4944ac28e764506bef0f943cb90)

SasSwart commented on agent/agentsocket/client.go:
```diff
@@ -0,0 +1,137 @@
+		Status:       resp.Status,
+		IsReady:      resp.IsReady,
+		Dependencies: dependencies,
+	}, nil
```
I'm going to keep it around for reasons stated [below](https://github.com/coder/coder/pull/20718#issuecomment-3573813577). It'll be easy to remove if it becomes a burden.

mafredri commented on agent/agentsocket/socket_unix.go:
```diff
@@ -1,9 +1,6 @@
-//go:build !windows
-
```
This change has me a bit confused. Why did we initially add not-implemented methods on Windows, and then remove it? Behavior when trying to use this on Windows is now somewhat undefined AFAICT.

mafredri reviewed (commented):
This is now basically in approve territory, but I'm on the fence about the Windows change. (And one more suggestion.)

SasSwart commented on agent/agentsocket/socket_windows.go:
```diff
@@ -1,33 +0,0 @@
-//go:build windows
```
We added it because we thought it was the best way to implement this at the time. We removed it because as we simplified code throughout this PR we realized that there was essentially nothing left of use in this file.

There is no reason to wait until createSocket is called to return an error that states that windows is not supported. It should happen as early as possible, where the client or server is initialised.

SasSwart replied:
We do still need the stubs for windows though. You're correct. Adding back.

SasSwart committed:
add a test to ensure that workspace agents can still initialize on windows despite the unsupported agentsocket feature

mafredri reviewed (approved):
Approved based on changes in next PR.

SasSwart committed:
feat(agent): add CLI commands to enable coder script interdependency (#20719)

relates to: https://github.com/coder/internal/issues/1094

This is number 5 of 5 pull requests in an effort to add agent script
ordering. It adds CLI commands that allow coder scripts to manage
dependencies between one another at runtime.

I used an LLM to produce some of these changes, but I have conducted
thorough self review and consider this contribution to be ready for an
external reviewer.

---------

Co-authored-by: Mathias Fredriksson <mafredri@gmail.com>
